File: 2015/lua/6/main-part-1.lua
local function get_action(line)
    local words = {}

    for word in line:gmatch("%w+") do
        table.insert(words, word)
    end

    if words[1] == "turn" then
        local action = words[2]
        local start_x = words[3]
        local start_y = words[4]
        local end_x = words[6]
        local end_y = words[7]
        return action, start_x, start_y, end_x, end_y
    elseif words[1] == "toggle" then
        local action = words[1]
        local start_x = words[2]
        local start_y = words[3]
        local end_x = words[5]
        local end_y = words[6]
        return action, start_x, start_y, end_x, end_y
    else
        print("Error")
    end
end

local function main()
    if #arg < 1 then
        print("Usage: lua main-part1.lua <input.txt>")
        os.exit(1)
    end

    local filename = arg[1]

    local file = io.open(filename, "r")

    if file == nil then
        print("Error: could not open file")
        os.exit(1)
    end

    local data = file:read("*all")

    local grid = {}
    for x = 0, 999 do
        grid[x] = {}
        for y = 0, 999 do
            grid[x][y] = false
        end
    end


    for line in data:gmatch("[^\n]+") do
        local action, startx, starty, endx, endy = get_action(line)

        for x = startx, endx do
            for y = starty, endy do
                if action == "on" then
                    grid[x][y] = true
                elseif action == "off" then
                    grid[x][y] = false
                elseif action == "toggle" then
                    grid[x][y] = not grid[x][y]
                end
            end
        end
    end

    local count = 0
    for x = 0, 999 do
        for y = 0, 999 do
            if grid[x][y] then
                count = count + 1
            end
        end
    end

    print("Part 1: " .. count)
end

main()

File: 2015/ruby/6/main-part-1.rb
#!/usr/bin/env ruby

def get_action(line)
    words = line.split(" ")
    if words[0] == "turn"
        action = words[1]
        start = words[2].split(",").map { |x| x.to_i }
        end_ = words[4].split(",").map { |x| x.to_i }
        return action, start, end_
    else
        action = words[0]
        start = words[1].split(",").map { |x| x.to_i }
        end_ = words[3].split(",").map { |x| x.to_i }
        return action, start, end_
    end
end

def main
    args = ARGV

    if args.length < 1
        puts "Usage: ruby main-part1.rb <input.txt>"
        exit 1
    end

    filename = args[0]
    file = File.open(filename, "r")

    if file.nil?
        puts "Could not open file"
        exit 1
    end

    lines = file.readlines
    file.close

    grid = Array.new(1000) { Array.new(1000, false) }
    
    lines.each do |line|
        action, start, end_ = get_action(line)
        for i in start[0]..end_[0]
            for j in start[1]..end_[1]
                if action == "on"
                    grid[i][j] = true
                elsif action == "off"
                    grid[i][j] = false
                else
                    grid[i][j] = !grid[i][j]
                end
            end
        end
    end

    count = 0
    for i in 0..999
        for j in 0..999
            if grid[i][j] == true
                count += 1
            end
        end
    end

    puts "Part 1: #{count}"
end

main()